File: lab03/src/main/java/com/example/lab03/ui/bmi/BmiFragment.java
package com.example.lab03.ui.bmi;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Base64;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebView;
import android.widget.EditText;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import com.example.lab03.databinding.FragmentBmiBinding;
import java.text.NumberFormat;

public class BmiFragment extends Fragment {

    // number formatter
    private static final NumberFormat numberFormat =
            NumberFormat.getNumberInstance();

    private int weight = 0; // weight entered by the user
    private double height = 0.0; // height entered by the user
    private TextView weightTextView; // shows formatted weight
    private TextView heightTextView; // shows formatted height
    private TextView bmiTextView; // shows calculated BMI
    private WebView bmiWebView; // BMI WebView Chart
    private final String unEncodedHtml =
            "  <html>\n" +
                    "  <head>\n" +
                    "    <script type=\"text/javascript\" src=\"https://www.gstatic.com/charts/loader.js\"></script>\n" +
                    "    <script type=\"text/javascript\">\n" +
                    "      google.charts.load('current', {'packages':['corechart']});\n" +
                    "      google.charts.setOnLoadCallback(drawChart);\n" +
                    "\n" +
                    "      function drawChart() {\n" +
                    "        var data = google.visualization.arrayToDataTable([\n" +
                    "          ['Month',    'Target BMI', 'Your BMI'],\n" +
                    "          ['October',   24.34,        25.02],\n" +
                    "          ['November',  24.34,        24.87],\n" +
                    "          ['December',  24.34,        24.90],\n" +
                    "          ['January',   24.34,        24.66],\n" +
                    "          ['February',  24.34,        24.46],\n" +
                    "          ['March',     24.34,        24.40]\n" +
                    "        ]);\n" +
                    "\n" +
                    "        var options = {\n" +
                    "          title: 'BMI over last 6 months',\n" +
                    "          curveType: 'function',\n" +
                    "          legend: { position: 'bottom' }\n" +
                    "        };\n" +
                    "\n" +
                    "        var chart = new google.visualization.LineChart(document.getElementById('curve_chart'));\n" +
                    "\n" +
                    "        chart.draw(data, options);\n" +
                    "      }\n" +
                    "    </script>\n" +
                    "  </head>\n" +
                    "  <body>\n" +
                    "    <div id=\"curve_chart\" style=\"width: 420px; height: 250px\"></div>\n" +
                    "  </body>\n" +
                    "</html>\n" +
                    "\n";

    private final String encodedHtml = Base64.encodeToString(unEncodedHtml.getBytes(), Base64.NO_PADDING);

    private FragmentBmiBinding binding;

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        BmiViewModel bmiViewModel =
                new ViewModelProvider(this).get(BmiViewModel.class);

        binding = FragmentBmiBinding.inflate(inflater, container, false);
        View root = binding.getRoot();

        weightTextView = binding.weightTextView;
        heightTextView = binding.heightTextView;
        bmiTextView = binding.bmiTextView;
        bmiWebView = binding.bmiWebView;
        bmiTextView.setText(numberFormat.format(0));

        bmiWebView.getSettings().setJavaScriptEnabled(true);
        bmiWebView.loadData(encodedHtml, "text/html", "base64");

        final EditText weightEditText = binding.weightEditText;
        weightEditText.addTextChangedListener(weightEditTextWatcher);

        final EditText heightEditText = binding.heightEditText;
        heightEditText.addTextChangedListener(heightEditTextWatcher);

        return root;
    }

    // calculate and display BMI
    private void calculate() {
        // calculate the BMI
        double bmi = weight / (height * height);

        // display BMI formatted as number with two fraction digits
        numberFormat.setMaximumFractionDigits(2);
        bmiTextView.setText(numberFormat.format(bmi));
    }

    // listener object for the weight EditText's text-changed events
    private final TextWatcher weightEditTextWatcher = new TextWatcher() {
        // called when the user modifies the weight
        @Override
        public void onTextChanged(CharSequence s, int start,
                                  int before, int count) {

            try { // get weight and display integer formatted value
                weight = (int) (Double.parseDouble(s.toString()));
                weightTextView.setText(numberFormat.format(weight));
            } catch (NumberFormatException e) { // if s is empty or non-numeric
                weightTextView.setText("");
                weight = 0;
            }

            calculate(); // update the BMI TextView
        }


        @Override
        public void afterTextChanged(Editable s) {
        }

        @Override
        public void beforeTextChanged(
                CharSequence s, int start, int count, int after) {
        }
    };

    // listener object for the EditText's text-changed events
    private final TextWatcher heightEditTextWatcher = new TextWatcher() {
        // called when the user modifies the height
        @Override
        public void onTextChanged(CharSequence s, int start,
                                  int before, int count) {

            try { // get height and display double formatted value
                height = Double.parseDouble(s.toString()) / 100.0;
                heightTextView.setText(numberFormat.format(height));
            } catch (NumberFormatException e) { // if s is empty or non-numeric
                heightTextView.setText("");
                height = 0.0;
            }

            calculate(); // update the BMI TextView
        }


        @Override
        public void afterTextChanged(Editable s) {
        }

        @Override
        public void beforeTextChanged(
                CharSequence s, int start, int count, int after) {
        }
    };

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: lab04/src/main/java/com/example/lab04/ui/bmi/BmiFragment.kt
package com.example.lab04.ui.bmi

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Base64
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.WebView
import android.widget.EditText
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import com.example.lab04.databinding.FragmentBmiBinding
import java.text.NumberFormat

class BmiFragment : Fragment() {

    private var _binding: FragmentBmiBinding? = null

    // This property is only valid between onCreateView and
    // onDestroyView.
    private val binding get() = _binding!!

    private val numberFormat = NumberFormat.getNumberInstance()
    private var weight = 0 // weight entered by the user
    private var height = 0.0 // height entered by the user
    private lateinit var weightTextView : TextView // shows formatted weight
    private lateinit var heightTextView : TextView // shows formatted height
    private lateinit var bmiTextView : TextView // shows calculated BMI

    private val unEncodedHtml = """  
<html>
  <head>
    <script type="text/javascript" src="https://www.gstatic.com/charts/loader.js"></script>
    <script type="text/javascript">
      google.charts.load('current', {'packages':['corechart']});
      google.charts.setOnLoadCallback(drawChart);

      function drawChart() {
        var data = google.visualization.arrayToDataTable([
          ['Month',    'Target BMI', 'Your BMI'],
          ['October',   24.34,        25.02],
          ['November',  24.34,        24.87],
          ['December',  24.34,        24.90],
          ['January',   24.34,        24.66],
          ['February',  24.34,        24.46],
          ['March',     24.34,        24.40]
        ]);

        var options = {
          title: 'BMI over last 6 months',
          curveType: 'function',
          legend: { position: 'bottom' }
        };

        var chart = new google.visualization.LineChart(document.getElementById('curve_chart'));

        chart.draw(data, options);
      }
    </script>
  </head>
  <body>
    <div id="curve_chart" style="width: 420px; height: 250px"></div>
  </body>
</html>
"""

    private val encodedHtml = Base64.encodeToString(unEncodedHtml.toByteArray(), Base64.NO_PADDING)

    override fun onCreateView(
            inflater: LayoutInflater,
            container: ViewGroup?,
            savedInstanceState: Bundle?
    ): View {
        _binding = FragmentBmiBinding.inflate(inflater, container, false)
        val root: View = binding.root

        weightTextView = binding.weightTextView
        heightTextView = binding.heightTextView
        bmiTextView = binding.bmiTextView
        bmiTextView.text = numberFormat.format(0)

        val bmiWebView: WebView = binding.bmiWebView
        bmiWebView.settings.javaScriptEnabled = true
        bmiWebView.loadData(encodedHtml, "text/html", "base64")

        val weightEditText: EditText = binding.weightEditText
        weightEditText.addTextChangedListener(weightEditTextWatcher)

        val heightEditText: EditText = binding.heightEditText
        heightEditText.addTextChangedListener(heightEditTextWatcher)

        return root
    }

    // calculate and display BMI
    private fun calculate() {
        // calculate the BMI
        val bmi = weight / (height * height)

        // display BMI formatted as number with two fraction digits
        numberFormat.maximumFractionDigits = 2
        bmiTextView.text = numberFormat.format(bmi)
    }

    // listener object for the weight EditText's text-changed events
    private val weightEditTextWatcher: TextWatcher = object : TextWatcher {
        // called when the user modifies the weight
        override fun onTextChanged(
            s: CharSequence, start: Int,
            before: Int, count: Int
        ) {
            try { // get weight and display integer formatted value
                weight = s.toString().toDouble().toInt()
                weightTextView.text = numberFormat.format(weight.toLong())
            } catch (e: NumberFormatException) { // if s is empty or non-numeric
                weightTextView.text = ""
                weight = 0
            }
            calculate() // update the BMI TextView
        }

        override fun afterTextChanged(s: Editable) {}
        override fun beforeTextChanged(
            s: CharSequence, start: Int, count: Int, after: Int
        ) {
        }
    }

    // listener object for the EditText's text-changed events
    private val heightEditTextWatcher: TextWatcher = object : TextWatcher {
        // called when the user modifies the height
        override fun onTextChanged(
            s: CharSequence, start: Int,
            before: Int, count: Int
        ) {
            try { // get height and display double formatted value
                height = s.toString().toDouble() / 100.0
                heightTextView.text = numberFormat.format(height)
            } catch (e: NumberFormatException) { // if s is empty or non-numeric
                heightTextView.text = ""
                height = 0.0
            }
            calculate() // update the BMI TextView
        }

        override fun afterTextChanged(s: Editable) {}
        override fun beforeTextChanged(
            s: CharSequence, start: Int, count: Int, after: Int
        ) {
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}